java/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/OnlineRetriever.java

```diff
@@ -55,22 +59,47 @@ public OnlineRetriever(
   public List<List<Feature>> getOnlineFeatures(
       List<Map<String, ValueProto.Value>> entityRows,
       List<ServingAPIProto.FeatureReferenceV2> featureReferences,
-      List<String> entityNames) {
+      List<String> entityNames,
+      Map<String, List<String>> entityNamesPerFeatureView) {
+
+    List<String> featureViewNames =
+        featureReferences.stream()
+            .map(FeatureReferenceV2::getFeatureViewName)
+            .distinct()
+            .collect(Collectors.toList());
+
+    Set<List<String>> uniqueEntityKeyCombinations =
+        featureViewNames.stream().map(entityNamesPerFeatureView::get).collect(Collectors.toSet());
+
+    List<List<RedisProto.RedisKeyV2>> redisKeys =
+        entityRows.stream()
+            .map(
+                entityRow ->
+                    uniqueEntityKeyCombinations.stream()
+                        .map(
+                            entitiesToRetain -> {
+                              Map<String, Value> res = new HashMap<>(entityRow);
+                              res.keySet().retainAll(entitiesToRetain);
+                              return res;
+                            })
+                        .collect(Collectors.toList()))
+            .map(es -> RedisKeyGenerator.buildRedisKeys(this.project, es))
```

Comment:
I suggest replace this to flatMap. So it will be flat `List<RedisProto.RedisKeyV2>` not list of lists. It seems that we know in advance how many sub keys for each row, our N rows will always transform to E * N keys, where E is amount of entities. So it will be easy to transform flat list of responses back into N element list

Reply:
I was under the impression that `getOnlineFeatures` assumes `featureReferences` and `entities` are all from the same feature view. The grouping features by feature view is done at the service level. This way this logic can be moved to the service layer and we don't have to handle it at each connector level, especially if there is a need to add a customized connector.